File: src/Rido.BFLite.Teams/Schema/TeamsChannelData.cs
using Rido.BFLite.Core.Schema;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Rido.BFLite.Teams.Schema;

public class TeamsChannelData : ChannelData
{
    public TeamsChannelData()
    {
    }

    public TeamsChannelData(ChannelData cd)
    {
        if (cd is not null)
        {
            if (cd.Properties.TryGetValue("teamsChannelId", out var channelIdObj) && channelIdObj is JsonElement jeChannelId && jeChannelId.ValueKind == JsonValueKind.String)
            {
                TeamsChannelId = jeChannelId.GetString();
            }

            if (cd.Properties.TryGetValue("teamsChannelId", out var teamsChannelId) && teamsChannelId is JsonElement teamsChannelIdJE && teamsChannelIdJE.ValueKind == JsonValueKind.String)
            {
                TeamsChannelId = teamsChannelIdJE.GetString();
            }

            if (cd.Properties.TryGetValue("channel", out var channelObj) && channelObj is JsonElement channelObjJE && channelObjJE.ValueKind == JsonValueKind.Object)
            {
                Channel = JsonSerializer.Deserialize<TeamsChannel?>(channelObjJE.GetRawText());
            }

            if (cd.Properties.TryGetValue("tenant", out var tenantObj) && tenantObj is JsonElement je && je.ValueKind == JsonValueKind.Object)
            {
                Tenant = JsonSerializer.Deserialize<TeamsChannelDataTenant>(je.GetRawText())!;
            }
        }
    }


    [JsonPropertyName("settings")]
    public TeamsChannelDataSettings? Settings { get; set; }

    [JsonPropertyName("teamsChannelId")]
    public string? TeamsChannelId { get; set; }

    [JsonPropertyName("teamsTeamId")]
    public string? TeamsTeamId { get; set; }

    [JsonPropertyName("channel")]
    public TeamsChannel? Channel { get; set; }

    [JsonPropertyName("team")]
    public Team? Team { get; set; }

    [JsonPropertyName("tenant")]
    public TeamsChannelDataTenant? Tenant { get; set; }

}

File: go/bflite/teams/schema/teams_channel_data.go
package schema

import "encoding/json"

// TeamsChannelData contains Teams-specific channel data.
type TeamsChannelData struct {
	ClientActivityID string                    `json:"clientActivityID,omitempty"`
	TeamsChannelID   string                    `json:"teamsChannelId,omitempty"`
	TeamsTeamID      string                    `json:"teamsTeamId,omitempty"`
	Channel          *TeamsChannel             `json:"channel,omitempty"`
	Team             *Team                     `json:"team,omitempty"`
	Tenant           *TeamsChannelDataTenant   `json:"tenant,omitempty"`
	Settings         *TeamsChannelDataSettings `json:"settings,omitempty"`

	// ExtensionData stores unknown properties
	ExtensionData map[string]json.RawMessage `json:"-"`
}

// TeamsChannel represents a Teams channel.
type TeamsChannel struct {
	ID   string `json:"id,omitempty"`
	Name string `json:"name,omitempty"`
}

// Team represents a Teams team.
type Team struct {
	ID           string `json:"id,omitempty"`
	Name         string `json:"name,omitempty"`
	AadGroupID   string `json:"aadGroupId,omitempty"`
	TenantID     string `json:"tenantId,omitempty"`
	Type         string `json:"type,omitempty"`
	ChannelCount int    `json:"channelCount,omitempty"`
	MemberCount  int    `json:"memberCount,omitempty"`
}

// TeamsChannelDataTenant represents tenant information.
type TeamsChannelDataTenant struct {
	ID string `json:"id,omitempty"`
}

// TeamsChannelDataSettings contains settings data.
type TeamsChannelDataSettings struct {
	SelectedChannel *TeamsChannel `json:"selectedChannel,omitempty"`
}

// MarshalJSON implements custom JSON marshaling.
func (tcd *TeamsChannelData) MarshalJSON() ([]byte, error) {
	type Alias TeamsChannelData
	data, err := json.Marshal((*Alias)(tcd))
	if err != nil {
		return nil, err
	}

	if len(tcd.ExtensionData) == 0 {
		return data, nil
	}

	var m map[string]json.RawMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	for k, v := range tcd.ExtensionData {
		if _, exists := m[k]; !exists {
			m[k] = v
		}
	}
	return json.Marshal(m)
}

// UnmarshalJSON implements custom JSON unmarshaling.
func (tcd *TeamsChannelData) UnmarshalJSON(data []byte) error {
	type Alias TeamsChannelData
	aux := &struct {
		*Alias
	}{
		Alias: (*Alias)(tcd),
	}

	if err := json.Unmarshal(data, aux); err != nil {
		return err
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}

	knownFields := []string{
		"clientActivityID", "teamsChannelId", "teamsTeamId",
		"channel", "team", "tenant", "settings",
	}
	for _, field := range knownFields {
		delete(raw, field)
	}

	if len(raw) > 0 {
		tcd.ExtensionData = raw
	}

	return nil
}
